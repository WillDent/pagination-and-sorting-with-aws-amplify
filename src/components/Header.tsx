import * as React from "react";
import {
  Text,
  TouchableOpacity,
  TouchableOpacityProps,
  View,
  StyleProp,
  ViewStyle,
  TextStyle
} from "react-native-web";
import { useRouter } from "next/router";

import { colors } from "../theme";

type LinkProps = { href: string } & TouchableOpacityProps;
const Link: React.ComponentType<LinkProps> = props => (
  <TouchableOpacity {...props} />
);

export const Header = () => {
  const router = useRouter();
  // throw "here " + router;
  // return null;

  // if (router)
  return (
    <View style={styles.header} accessibilityLabel="Header Navigation">
      <Link
        style={[
          styles.headerLink,
          router.pathname === "/me" && styles.headerSelected
        ]}
        href="/me"
        accessible={true}
        onPress={() => {
          router.push("/me");
        }}
        accessibilityRole="link"
      >
        <Text style={styles.headerText}>My Profile</Text>
      </Link>

      <Link
        style={[
          styles.headerLink,
          (router.pathname === "/channels" || router.pathname === "/") &&
            styles.headerSelected
        ]}
        href="/channels"
        accessible={true}
        onPress={() => {
          router.push("/channels");
        }}
        accessibilityRole="link"
      >
        <Text style={styles.headerText}>Channels</Text>
      </Link>
    </View>
  );
};

const styles: { [key: string]: StyleProp<ViewStyle | TextStyle> } = {
  container: {
    width: "100%",
    height: "100%",
    backgroundColor: colors.primaryLight
  },
  header: {
    display: "flex",
    flexDirection: "row",
    height: "10%",
    backgroundColor: colors.primary
  },
  headerLink: { width: "50%", alignItems: "center", justifyContent: "center" },
  headerSelected: {
    borderBottomColor: colors.highlight,
    borderBottomWidth: 2
  },
  headerText: {
    color: "white",
    fontWeight: "800",
    fontSize: 20
  }
};
