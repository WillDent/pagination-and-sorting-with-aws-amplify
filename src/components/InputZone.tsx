import * as React from "react";
import { View, Text, TouchableOpacity, TextInput } from "react-native-web";

import { colors } from "../theme";
import { FlexSpacer } from "./utils";

export const InputZone = ({ onSubmit, placeholder, buttonText }) => {
  return (
    <View
      style={{
        height: "10%",
        backgroundColor: colors.primaryDark,
        display: "flex",
        flexDirection: "row",
        alignItems: "center"
      }}
    >
      <FlexSpacer />
      <TextInput
        placeholder={placeholder}
        style={{
          flex: 3,
          height: "50%",
          color: colors.primaryDark,
          padding: 10,
          borderRadius: 3,
          backgroundColor: "white"
        }}
        placeholderTextColor={colors.primaryOpaque(0.7)}
      />
      <FlexSpacer />
      <View
        style={{
          flex: 1,
          height: "50%"
        }}
      >
        <TouchableOpacity
          onPress={() => {
            console.warn("Creating new topic");
          }}
          style={{
            backgroundColor: colors.highlight,
            height: "100%",
            display: "flex",
            alignItems: "center",
            borderRadius: 15,
            justifyContent: "center"
          }}
        >
          <Text
            style={{
              color: "white",
              textAlign: "center"
            }}
          >
            {buttonText}
          </Text>
        </TouchableOpacity>
      </View>
      <FlexSpacer />
    </View>
  );
};
