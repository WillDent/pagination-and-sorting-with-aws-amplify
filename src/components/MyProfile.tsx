import * as React from "react";
import { View, TouchableOpacity, TextInput } from "react-native-web";

import { colors } from "../theme";
import { ImportantText, StaticSpacer, NormalText } from "./utils";

const textInputStyle = {
  backgroundColor: "white",
  color: colors.primaryDark,
  height: 30,
  padding: 20,
  borderRadius: 12
};

export const MyProfile = () => {
  return (
    <>
      <View
        style={{
          padding: 20,
          height: "75%"
        }}
      >
        <ImportantText accessibilityRole={"header"}>My profile</ImportantText>

        <StaticSpacer />
        <NormalText>Username</NormalText>
        <StaticSpacer />
        <TextInput style={textInputStyle} placeholder={"Username"} />

        <StaticSpacer />
        <NormalText>Bio</NormalText>
        <StaticSpacer />
        <TextInput style={textInputStyle} placeholder={"Bio"} />

        <StaticSpacer />
        <NormalText>Link</NormalText>
        <StaticSpacer />
        <TextInput
          style={textInputStyle}
          placeholder={"Twitter/Github/Medium/..."}
        />

        <StaticSpacer />
      </View>
      <View
        style={{
          padding: 20,
          height: "15%"
        }}
      >
        <TouchableOpacity
          style={{
            backgroundColor: colors.highlight,
            padding: 20
          }}
          onPress={() => {}}
        >
          <NormalText style={{ textAlign: "center" }}>Save</NormalText>
        </TouchableOpacity>
      </View>
    </>
  );
};
