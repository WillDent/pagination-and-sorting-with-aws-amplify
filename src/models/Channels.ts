import { API, graphqlOperation } from "aws-amplify";

import {
  createChannelList,
  createChannel as createChannelQuery
} from "../graphql/mutations";
import { State, ChannelType } from "../types";
import config from "../aws-exports.js";

API.configure(config);

type CustomChannelList = {
  data: {
    getChannelList: {
      id: string;
      channels: {
        items: {
          id: string;
          name: string;
          createdAt: string;
          updatedAt: string;
          messages: {
            items: {
              id: string;
              text: string;
              createdAt: string;
              senderId: string;
            }[];
          };
        }[];
        nextToken: string;
      };
    };
  };
};
// Just to get nice formatting for our queries
const gql = String.raw;

const getCustomChannelList = ({
  nextToken
}: {
  nextToken: string;
}) => gql`query GetChannelList($id: ID!) {
  getChannelList(id: $id) {
    id
    channels(limit: 5, sortDirection: DESC, ${
      nextToken !== "" ? `nextToken:"${nextToken}"` : ``
    }) {
      items {
        id
        name
        createdAt
        updatedAt
        messages {
          items {
            id
            createdAt
            senderId
            text
          } 
          
        }
      }
      nextToken
    }
  }
}
`;

export const getChannels = async (): Promise<State["channels"]> => {
  console.warn("Getting channels");

  const channelList = await (API.graphql(
    graphqlOperation(
      getCustomChannelList({
        nextToken: ""
        // "eyJ2ZXJzaW9uIjoxLCJ0b2tlbiI6IkFRSUNBSGg5OUIvN3BjWU41eE96NDZJMW5GeGM4WUNGeG1acmFOMUpqajZLWkFDQ25BRVJIQTJSUWNYM0g5ZWgwNDQyVnJTUEFBQUROVENDQXpFR0NTcUdTSWIzRFFFSEJxQ0NBeUl3Z2dNZUFnRUFNSUlERndZSktvWklodmNOQVFjQk1CNEdDV0NHU0FGbEF3UUJMakFSQkF3cWV1MElzRXF5bTE4L0lvTUNBUkNBZ2dMbzJnNDdoc1R4dE40bUY5K3VmSmZEOG1sQ3lnaXB4RGszanFDR1RCanRhT3FvcGd1WDZUOWxXYlpCZkkyY0kvU1FlUGs1Wmt1SnlaZ2lDNThnMTB2bEZaTlhINmdBSWpxNVFMeThnWGdMZnIvUEttRnhzTzlqeURrRTZRREVlNy81Nk5qSjRkYjh2bWVhaXAvY0VvWlhQc3BoZE9qYStPbkZiSFRZQ2tMREhTcG12dUxtODNKNy9PMnp5OFZTOUtaMkJEV2t6OHpqVE93UkFKKzRRUjRiN2RrR1BKcHJJeFBHa01OUzF5RmZvdnJqMndVTzViS3FYKzJlNW1lMExVOUZNVXNxTVYzYk9GUGwwcTFHTERycDRtTU0zUGdXM1I2dlVQUkpkTkZ3TFhjbW1KVjZad21id0JhV2tDcEhXUFJmYVJDQ0VGSDJSZjJCVk05QlI5cS9vMFpHUGpNRWVGYjlOYzRsYnFObUN4YVdLcjBaaGgySVB5MEk3RVlBNTF4ajN1TUZrZFZQZHp3R3U0SjJoVGtwZXRianNaSTd2M3UvR0hDVHhKc3Rla1FMSlhnelNDbU9acXdyQU1iMElmb0NNRE1KWmxqeUJ4R2dKS1Jabkh2OWRqcmFlTGFwVDR5ek5Ib2U4cDg2YVo2RkZrWm9QTHpUSWFuMjIzSXRBUXRKUzZGZTRqRG9kdGplT2RRRnV6V2loOWh0U09CSmlCR2Z2Z2grRi9VL1JVNUEzVlpOczRvRDlQWDBXVjhNZU9Yc1E5RkxDR1hOR0ZjQTZUSjBMR2NXMUZUbDJsazhSRm44QXJ0RFJYS1BESlNwZGVZU1ZGbUlmdmFPSm9ESWd4RmhYVVRvL2dwdmNHMVJrRERGbWFDdWtLTDJYY1U1SEpQc2xid1oyWmlwUkc0a0xLcEJWV3c4ekNXSDdwYnJWZ3RrM0cxSGhpUVdBUm5QQkwrcXM0cFM4MlJvVEM5Q0VvY0VQVjJRazZXWEhxeFVxM1ZiMkVucHRuVlRFYlNoZlFpZk1ybzNyQWpzQnlYNXBSQWdlWlN6cHZpREtMMTZvaCtUNzR3bXR1ZVRYb0ZmYUJhbmZGOElCZjEvUTYrdHdsWG1BZ2RzTzQ4MHJjR09QVkhKb3ZqSlIyMGRJVkVpQy9odTlQekxoblAweFNyUndXZjY4SlVMMUZWWGNscDQ4a0FwQUZXQ1JUTWRTSVJFNndJZFM3NmxaRVZSMm9TamJ4R2lOUzZyZ3VhNm1zeGFXZTZPN0JtZkp4VEJkVTlWQkZRbHphVFBFT20rZFNjVmhCd0VxYUFweEZtMmhFcnMxRG11In0="
      }),
      {
        id: "global"
      }
    )
  ) as Promise<CustomChannelList>);
  if (channelList.data.getChannelList === null) {
    try {
      await API.graphql(
        graphqlOperation(createChannelList, { input: { id: "global" } })
      );
    } catch (err) {
      console.warn("Failed to create channel list ", err);
    }
    return [];
  }
  console.warn(channelList.data.getChannelList);
  // Massage the data a bit to give it the shape of our local state
  let channels = channelList.data.getChannelList.channels.items;
  const channelsWithMessages = channels.map(channel => ({
    ...channel,
    messages: channel.messages.items
  }));
  return channelsWithMessages;
};

export const createChannel = async (channel: ChannelType) => {
  try {
    const input = {
      ...channel,
      messages: undefined,
      channelChannelListId: "global"
    };
    await API.graphql(graphqlOperation(createChannelQuery, { input }));
  } catch (err) {
    console.warn("Could not create channel", channel, err);
  }
};
