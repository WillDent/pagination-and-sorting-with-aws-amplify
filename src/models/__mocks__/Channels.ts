import API, { graphqlOperation } from "@aws-amplify/api";
// import PubSub from "@aws-amplify/pubsub";
import memoize from "lodash.memoize";

import {
  OnCreateChannelInListSubscription,
  OnUpdateChannelInListSubscription
} from "../../API";

import {
  createChannelList,
  createChannel as createChannelQuery
} from "../../graphql/mutations";
import {
  onCreateChannelInList,
  onUpdateChannelInList
} from "../../graphql/subscriptions";
import { State, ChannelType, List, Listener, MessageType } from "../../types";

//@ts-ignore
import config from "../../aws-exports.js";
import { getChannelList as getChannelListQuery } from "./custom-queries";

// API.configure(config);
// PubSub.configure(config);

type CustomChannelList = {
  data: {
    getChannelList: {
      id: string;
      channels: List<{
        id: string;
        name: string;
        createdAt: string;
        updatedAt: string;
        creatorId: string;
        messages: List<MessageType>;
      }>;
    };
  };
};

export let resolveGetChannels;

export const getGetChannelsResolver = () => resolveGetChannels;

export const getChannels = jest.fn();

export const createChannel = jest.fn();

export const onCreateChannel = jest.fn();

export const onUpdateChannel = jest.fn();
