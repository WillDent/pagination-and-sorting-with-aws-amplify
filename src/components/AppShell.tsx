import { View } from "react-native-web";
import * as React from "react";

import { Header } from "./Header";
import { colors } from "../theme";
import { State, Dispatcher } from "../types";
import { createUserIfNotExists, getUser } from "../models/User";
import { DispatcherContext } from "../state";

const AppShell: React.FC<{ state: State; dispatch: Dispatcher }> = ({
  state,
  dispatch,
  children
}) => {
  React.useEffect(() => {
    createUserIfNotExists(state.me);
  }, []);
  React.useEffect(() => {
    let isMounted = true;
    getUser(state.me.id).then(getUserResponse => {
      if (isMounted === false) return;
      let meFromServer = getUserResponse.data.getUser;
      if (meFromServer === null) return;
      dispatch({ type: "set-my-info", payload: meFromServer });
    });
    return () => {
      isMounted = false;
    };
  }, [state.me.id]);
  return (
    <DispatcherContext.Provider value={dispatch}>
      <View
        style={{
          width: "100%",
          height: "100%",
          backgroundColor: colors.primaryLight
        }}
      >
        <Header />
        {children}
      </View>
    </DispatcherContext.Provider>
  );
};

export default AppShell;
