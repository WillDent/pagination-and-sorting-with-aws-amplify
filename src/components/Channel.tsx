import nanoid from "nanoid";
import * as React from "react";
import { FlatList, Text, View } from "react-native-web";

import { colors } from "../theme";
import { MessageType, Dispatcher, State } from "../types";
import { InputZone } from "./InputZone";
import { createMessage, getChannelMessages } from "../models/Channel";
import { DispatcherContext } from "../state";

export const Message = ({ message }: { message: MessageType }) => {
  return (
    <View
      style={{
        padding: 20,
        backgroundColor: colors.primaryDark,
        marginBottom: 10,
        marginTop: 10,
        marginLeft: 30,
        marginRight: 30,
        display: "flex",
        flexDirection: "row",
        borderRadius: 12
      }}
    >
      <View style={{ flex: 10 }}>
        <Text style={{ color: "white", fontWeight: "bold", marginBottom: 5 }}>
          {message.text}
        </Text>
        <Text style={{ color: "white" }}>
          Sent by Username {message.senderId}
        </Text>
        <Text style={{ color: "white" }}>
          Created at: {new Date(Number(message.createdAt)).toLocaleString()}
        </Text>
      </View>
    </View>
  );
};

export const Channel = ({
  messages = { items: [], nextToken: "" },
  shouldScrollDown,
  channelId
}: {
  messages: State["channels"]["items"][0]["messages"];
  shouldScrollDown?: number;
  channelId: string;
}) => {
  const dispatch = React.useContext(DispatcherContext);
  const flatlistRef = React.useRef<null | FlatList<MessageType>>(null);
  React.useEffect(() => {
    if (flatlistRef.current === null) return;
    // flatlistRef.current.scrollToEnd();
  }, [shouldScrollDown]);
  return (
    <FlatList
      inverted={true}
      ref={flatlistRef}
      style={{
        height: "80%"
      }}
      keyExtractor={item => item.id}
      data={messages.items}
      renderItem={({ item }) => <Message key={item.id} message={item} />}
      onEndReached={() => {
        if (messages.nextToken === null) return;
        getChannelMessages(channelId, messages.nextToken).then(messages => {
          dispatch({
            type: "append-messages",
            payload: { channelId, messages }
          });
        });
      }}
      onEndReachedThreshold={0.01}
    />
  );
};

type ChannelRouteProps = {
  channelId: string;
  channels: State["channels"];
  me: State["me"];
};

export const ChannelRoute = ({
  channelId,
  channels,
  me
}: ChannelRouteProps) => {
  const dispatch = React.useContext(DispatcherContext);
  const [shouldScrollDown, setScrollDown] = React.useState(0);
  const channelIndex = channels.items.findIndex(
    channel => channel.id === channelId
  );

  React.useEffect(() => {
    getChannelMessages(channelId, "").then(messages => {
      dispatch({ type: "append-messages", payload: { channelId, messages } });
    });
  }, [channelId]);

  if (channelIndex === -1) {
    return <Text>Channel {channelId} doesn't exist yet</Text>;
  }
  const addMessage = (content: string, dispatch: Dispatcher) => {
    const message = {
      text: content,
      createdAt: `${Date.now()}`,
      id: nanoid(),
      senderId: me.id,
      messageChannelId: channelId
    };
    dispatch({
      type: "prepend-message",
      payload: message
    });
    setScrollDown(Date.now());
    createMessage(message);
  };

  return (
    <>
      <Channel
        messages={channels.items[channelIndex].messages}
        shouldScrollDown={shouldScrollDown}
        channelId={channelId}
      />
      <InputZone
        placeholder={"Create a new message"}
        onSubmit={content => {
          addMessage(content, dispatch);
        }}
        buttonText={"Send message"}
      />
    </>
  );
};
