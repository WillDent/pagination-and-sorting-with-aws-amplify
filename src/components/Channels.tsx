import Head from "next/head";
import { useRouter } from "next/router";
import * as React from "react";
import {
  ActivityIndicator,
  FlatList,
  Text,
  TouchableOpacity,
  View
} from "react-native-web";

import { getActions } from "../actions";
import { onCreateMessage } from "../models/Channel";
import {
  getChannels,
  onCreateChannel,
  onUpdateChannel
} from "../models/Channels";
import { DispatcherContext, useAppReducer } from "../state";
import { colors } from "../theme";
import { ChannelType, State } from "../types";
import AppShell from "./AppShell";
import { InputZone } from "./InputZone";

type Props = { channel: ChannelType; me: State["me"] };

const ChannelCard = (props: Props) => {
  const dispatch = React.useContext(DispatcherContext);
  const {
    channel,
    me: { id: myId }
  } = props;
  const { messages = { items: [] }, id: channelId } = channel;
  const router = useRouter();
  const lastMessage = messages.items.length > 0 ? messages.items[0].text : "";
  React.useEffect(() => {
    const subscription = onCreateMessage(channelId).subscribe(
      data => {
        const newMessage = data.value.data.onCreateMessageInChannel;
        if (newMessage === null) {
          return;
        }
        const senderId = newMessage.senderId;
        if (senderId === myId) {
          return;
        }
        //@ts-ignore
        dispatch({ type: "prepend-message", payload: newMessage });
      },
      () => {
        console.error("Error onCreateMessage");
      }
    );
    return () => {
      subscription.unsubscribe();
    };
  }, [channelId]);
  return (
    <TouchableOpacity
      style={{
        padding: 20,
        backgroundColor: colors.primaryDark,
        borderColor: colors.highlightOpaque(0.3),
        borderWidth: 2,
        marginBottom: 10,
        marginTop: 10,
        marginLeft: 30,
        marginRight: 30,
        display: "flex",
        flexDirection: "row",
        borderRadius: 12,
        shadowColor: "#000",
        shadowOffset: {
          width: 0,
          height: 7
        },
        shadowOpacity: 0.43,
        shadowRadius: 9.51,

        elevation: 10
      }}
      accessibilityRole="link"
      {...{
        // Being dishonest with typescript because of the lack of react-native-web types
        href: `/channel/${1}`
      }}
      onPress={() => {
        router.push(`/channel?id=${channel.id}`);
      }}
    >
      <View style={{ flex: 10 }}>
        <View style={{ display: "flex", flexDirection: "row" }}>
          <Text style={{ color: "white" }}>Channel name: </Text>
          <Text style={{ color: "white", fontWeight: "bold", marginBottom: 5 }}>
            {channel.name}
          </Text>
        </View>
        <Text
          style={{
            color: "white",
            // fontWeight: "bold",
            marginTop: 5,
            marginBottom: 5
          }}
        >
          Last updated: {new Date(Number(channel.updatedAt)).toLocaleString()}
        </Text>
        <Text style={{ color: "white", marginTop: 15 }}>{lastMessage}</Text>
      </View>
    </TouchableOpacity>
  );
};

export const Channels = ({
  channels,
  me
}: {
  channels: State["channels"];
  me: State["me"];
}) => {
  const dispatch = React.useContext(DispatcherContext);
  const flatlistRef = React.useRef<null | FlatList<ChannelType>>(null);
  const [isLoadingPosition, setIsLoadingPosition] = React.useState<
    "none" | "top" | "bottom"
  >("none");
  React.useEffect(() => {
    let isMounted = true;
    setIsLoadingPosition("top");
    const onCreateSubscription = onCreateChannel().subscribe(
      response => {
        const channel = response.value.data.onCreateChannelInList;
        if (channel === null || channel.creatorId === me.id) return;
        dispatch({
          type: "prepend-channel",
          payload: {
            id: channel.id,
            name: channel.name,
            createdAt: channel.createdAt,
            updatedAt: channel.updatedAt,
            creatorId: channel.creatorId || "",
            messages: {
              items: [],
              nextToken: ""
            }
          }
        });
      },
      err => {
        console.warn("onCreateChannel subscription error ", err);
      }
    );

    const onUpdateChannelSubscription = onUpdateChannel().subscribe(
      response => {
        const channel = response.value.data.onUpdateChannelInList;
        if (channel === null) return;
        dispatch({
          type: "update-channel",
          payload: { ...channel, creatorId: channel.creatorId || "" }
        });
        dispatch({
          type: "move-to-front",
          payload: { channelId: channel.id }
        });
      },
      err => {
        console.error("Error onUpdateChannelSubscription", err);
      }
    );

    getChannels()
      .then(channels => {
        if (!isMounted) return;
        setIsLoadingPosition("none");
        dispatch({ type: "set-channels", payload: channels });
      })
      .catch(err => {
        console.warn("Error fetching channels ", err);
      });

    return () => {
      isMounted = false;
      onCreateSubscription.unsubscribe();
      onUpdateChannelSubscription.unsubscribe();
    };
  }, []);
  return (
    <>
      <FlatList
        inverted={false}
        style={{
          height: "80%"
        }}
        ListHeaderComponent={() => (
          <View style={{ height: 30 }}>
            {isLoadingPosition === "top" && (
              <ActivityIndicator
                style={{ height: 30 }}
                animating={true}
                color={colors.highlight}
              />
            )}
          </View>
        )}
        ListFooterComponent={() => (
          <View style={{ height: 30 }}>
            {isLoadingPosition === "bottom" && (
              <ActivityIndicator
                style={{ height: 30 }}
                animating={true}
                color={colors.highlight}
              />
            )}
          </View>
        )}
        ref={flatlistRef}
        keyExtractor={item => {
          return item.id;
        }}
        data={channels.items}
        renderItem={({ item: channel }) => (
          <ChannelCard channel={channel} me={me} />
        )}
        onEndReached={() => {
          if (channels.nextToken === null) return;
          setIsLoadingPosition("bottom");
          getChannels(channels.nextToken).then(nextChannels => {
            setIsLoadingPosition("none");
            dispatch({ type: "append-channels", payload: nextChannels });
          });
        }}
        onEndReachedThreshold={0.1}
      />
    </>
  );
};

export const ChannelsRoute = () => {
  const [state, dispatch] = useAppReducer();
  const actions = getActions(dispatch);
  return (
    <>
      <Head>
        <title>Channels</title>
      </Head>
      <AppShell state={state} dispatch={dispatch}>
        <Channels channels={state.channels} me={state.me} />
        <InputZone
          placeholder={"Create a new channel"}
          onSubmit={content => {
            actions.addChannel(content, state.me.id);
          }}
          buttonText={"Create channel"}
        />
      </AppShell>
    </>
  );
};
