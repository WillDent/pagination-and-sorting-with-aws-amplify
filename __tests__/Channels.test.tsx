import * as React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import { RouterContext } from "next-server/dist/lib/router-context";

import "@testing-library/jest-dom/extend-expect";

import { NextRouter } from "next/router";
import { ChannelsRoute } from "../src/components/Channels";

describe("Channels route", () => {
  test("exports", () => {
    expect(ChannelsRoute).toBeDefined();
  });
  test("renders", () => {
    const router = ({
      pathname: "/",
      route: "/",
      query: {},
      asPath: "/",
      push: jest.fn()
    } as unknown) as NextRouter;
    const { container, getByLabelText } = render(
      <RouterContext.Provider value={router}>
        <ChannelsRoute />
      </RouterContext.Provider>
    );
    // Temp
    expect(container).toMatchSnapshot();
    const createNewChannelInput = getByLabelText("Create a new channel");
    const createNewChannelSubmitButton = getByLabelText("Create channel");
    expect(createNewChannelInput).toBeVisible();
    expect(createNewChannelSubmitButton).toBeVisible();
    fireEvent.change(createNewChannelInput, {
      target: {
        value: "Test Channel"
      }
    });
    fireEvent.click(createNewChannelSubmitButton);
  });
});
export default "";
