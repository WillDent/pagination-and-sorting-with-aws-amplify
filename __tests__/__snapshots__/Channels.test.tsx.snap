// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Channels route renders 1`] = `
<div>
  <div
    class="css-view-1dbjc4n"
    style="background-color: rgb(27, 37, 77); height: 100%; width: 100%;"
  >
    <h1
      aria-level="1"
      class="css-reset-4rbku5 css-view-1dbjc4n"
      role="heading"
      style="background-color: rgb(18, 25, 50); display: flex; flex-direction: row; height: 10%;"
    >
      <a
        class="css-reset-4rbku5 css-cursor-18t94o4 css-view-1dbjc4n r-cursor-1loqt21 r-touchAction-1otgn73 r-transitionDuration-eafdt9 r-transitionProperty-1i6wzkk r-userSelect-lrvibr"
        data-focusable="true"
        href="/me"
        role="link"
        style="align-items: center; justify-content: center; width: 50%;"
      >
        <div
          class="css-text-901oao"
          dir="auto"
          style="color: rgb(255, 255, 255); font-size: 20px; font-weight: 800;"
        >
          My Profile
        </div>
      </a>
      <a
        class="css-reset-4rbku5 css-cursor-18t94o4 css-view-1dbjc4n r-cursor-1loqt21 r-touchAction-1otgn73 r-transitionDuration-eafdt9 r-transitionProperty-1i6wzkk r-userSelect-lrvibr"
        data-focusable="true"
        href="/channels"
        role="link"
        style="align-items: center; border-bottom-width: 2px; justify-content: center; width: 50%;"
      >
        <div
          class="css-text-901oao"
          dir="auto"
          style="color: rgb(255, 255, 255); font-size: 20px; font-weight: 800;"
        >
          Channels
        </div>
      </a>
    </h1>
    <main
      style="width: 100%; height: 80%;"
    >
      <div
        class="css-view-1dbjc4n r-WebkitOverflowScrolling-150rngu r-flexDirection-eqz5dr r-flexGrow-16y2uox r-flexShrink-1wbh5a2 r-overflowX-11yh6sk r-overflowY-1rnoaur r-transform-1sncvnh"
        style="height: 100%;"
      >
        <div
          class="css-view-1dbjc4n"
        >
          <div
            class="css-view-1dbjc4n"
          >
            <div
              class="css-view-1dbjc4n"
              style="height: 30px;"
            >
              <div
                aria-valuemax="1"
                aria-valuemin="0"
                class="css-view-1dbjc4n r-alignItems-1awozwy r-justifyContent-1777fci"
                role="progressbar"
                style="height: 30px;"
              >
                <div
                  class="css-view-1dbjc4n r-animationDuration-17bb2tj r-animationIterationCount-1muvv40 r-animationKeyframes-127358a r-animationTimingFunction-1ldzwu0 r-height-z80fyv r-width-19wmn03"
                >
                  <svg
                    height="100%"
                    viewBox="0 0 32 32"
                    width="100%"
                  >
                    <circle
                      cx="16"
                      cy="16"
                      fill="none"
                      r="14"
                      stroke-width="4"
                      style="stroke: #61dafb; opacity: 0.2;"
                    />
                    <circle
                      cx="16"
                      cy="16"
                      fill="none"
                      r="14"
                      stroke-width="4"
                      style="stroke: #61dafb; stroke-dasharray: 80; stroke-dashoffset: 60;"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </main>
    <div
      class="css-view-1dbjc4n"
      style="align-items: center; background-color: rgb(14, 20, 43); display: flex; flex-direction: row; height: 10%;"
    >
      <div
        class="css-view-1dbjc4n"
        style="flex-grow: 0.1; flex-shrink: 1; flex-basis: 0%;"
      />
      <input
        aria-label="Create a new channel"
        autocapitalize="sentences"
        autocomplete="on"
        autocorrect="on"
        class="css-textinput-1cwyjr8 r-placeholderTextColor-1877jhj"
        data-focusable="true"
        dir="auto"
        placeholder="Create a new channel"
        spellcheck="true"
        style="background-color: rgb(255, 255, 255); border-top-left-radius: 3px; border-top-right-radius: 3px; border-bottom-right-radius: 3px; border-bottom-left-radius: 3px; color: rgb(14, 20, 43); flex-grow: 3; flex-shrink: 1; flex-basis: 0%; height: 50%; padding: 10px 10px 10px 10px;"
        type="text"
        value=""
      />
      <div
        class="css-view-1dbjc4n"
        style="flex-grow: 0.1; flex-shrink: 1; flex-basis: 0%;"
      />
      <div
        class="css-view-1dbjc4n"
        style="flex-grow: 1; flex-shrink: 1; flex-basis: 0%; height: 50%;"
      >
        <div
          aria-label="Create channel"
          class="css-view-1dbjc4n r-cursor-1loqt21 r-touchAction-1otgn73 r-transitionDuration-eafdt9 r-transitionProperty-1i6wzkk r-userSelect-lrvibr"
          data-focusable="true"
          style="align-items: center; background-color: rgba(97, 218, 251, 0.3); border-top-left-radius: 15px; border-top-right-radius: 15px; border-bottom-right-radius: 15px; border-bottom-left-radius: 15px; display: flex; height: 100%; justify-content: center;"
          tabindex="0"
        >
          <div
            class="css-text-901oao"
            dir="auto"
            style="color: rgb(255, 255, 255); text-align: center;"
          >
            Create channel
          </div>
        </div>
      </div>
      <div
        class="css-view-1dbjc4n"
        style="flex-grow: 0.1; flex-shrink: 1; flex-basis: 0%;"
      />
    </div>
  </div>
</div>
`;
